docs/core/testing/unit-testing-mstest-writing-tests.md

```diff
@@ -0,0 +1,396 @@
+ms.date: 07/18/2024
+---
+
+# Writing tests with MSTest
+
+In this article, you will learn about the APIs and conventions used by MSTest to help you write and shape your tests.
+
+## Attributes
+
+MSTest uses custom attributes to identify and customize tests.
+
+To help provide a clearer overview of the testing framework, this section organizes the members of the <xref:Microsoft.VisualStudio.TestTools.UnitTesting> namespace into groups of related functionality.
+
+> [!NOTE]
+> Attribute elements, whose names end with "Attribute", can be used either with or without "Attribute" on the end and for parameterless constructors with or without parenthesis. For example, the following code examples function identically:
+>
+> `[TestClass()]`
+>
+> `[TestClassAttribute()]`
+>
+> `[TestClass]`
+>
+> `[TestClassAttribute]`
+
+### Attributes used to identify test classes and methods
+
+Every test class must have the `TestClass` attribute, and every test method must have the `TestMethod` attribute. For more information, see [Anatomy of a unit test](https://learn.microsoft.com/previous-versions/ms182517(v=vs.110)).
+
+#### TestClassAttribute
+
+The [TestClass](<xref:Microsoft.VisualStudio.TestTools.UnitTesting.TestClassAttribute>) attribute marks a class that contains tests and, optionally, initialize or cleanup methods.
+
+This attribute can be extended to update or extend the behavior.
+
+Example:
+
+```csharp
+[TestClass]
+public class MyTestClass
+{    
+}
+```
+
+#### TestMethodAttribute
+
+The [TestMethod](<xref:Microsoft.VisualStudio.TestTools.UnitTesting.TestMethodAttribute>) attribute is used inside a `TestClass` to define the actual test method to run.
+
+The method should be an instance method defined as `public void` or `public Task` (optionally `async`) and be parameterless.
+
+**Example**
+
+```csharp
+[TestClass]
+public class MyTestClass
+{
+    [TestMethod]
+    public void TestMethod()
+    {
+    }
+}
+```
+
+```csharp
+[TestClass]
+public class MyTestClass
+{
+    [TestMethod]
+    public async Task TestMethod()
+    {
+    }
+}
+```
+
+### Attributes used for data-driven testing
+
+Use the following elements to set up data-driven tests. For more information, see [Create a data-driven unit test](/visualstudio/test/how-to-create-a-data-driven-unit-test.md) and [Use a configuration file to define a data source](/visualstudio/test/walkthrough-using-a-configuration-file-to-define-a-data-source.md).
+
+- <xref:Microsoft.VisualStudio.TestTools.UnitTesting.DataTestMethodAttribute>
+
+- <xref:Microsoft.VisualStudio.TestTools.UnitTesting.DataRowAttribute>
+
+- <xref:Microsoft.VisualStudio.TestTools.UnitTesting.DynamicDataAttribute>
+
+- <xref:Microsoft.VisualStudio.TestTools.UnitTesting.DataSourceAttribute>
+
+#### DataRow
+
+The `DataRowAttribute` allows you to provide inline data used when invoking the test method. It can appear one or multiple times on a test method. It should be combined with `TestMethodAttribute` or `DataTestMethodAttribute`.
```

Comment:
```suggestion
The `DataRowAttribute` allows you to run the same test method with multiple different inputs. It can appear one or multiple times on a test method. It should be combined with `TestMethodAttribute` or `DataTestMethodAttribute`.
```